src/identity/customca.rs

```diff
@@ -0,0 +1,141 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::CertificateProvider;
+use async_trait::async_trait;
+use kube::{
+    api::{Api, PostParams, ListParams, WatchEvent},
+    client::Client,
+};
+use tracing::{instrument, warn};
+use crate::{identity::manager::Identity, tls::SanChecker};
+use crate::identity::Error;
+use crate::tls;
+use futures::{StreamExt, TryStreamExt};
+use k8s_openapi::{api::certificates::v1::{CertificateSigningRequest, CertificateSigningRequestSpec}, ByteString};
+use std::str;
+use rand::prelude::*;
+
+#[derive(Clone)]
+pub struct CustomSigner {
+    pub signer: Option<std::string::String>,
+    pub client: Client,
+}
+
+impl CustomSigner {
+    pub fn new(signer: Option<std::string::String>, client: Client) -> Result<CustomSigner, Error> {
+        Ok(CustomSigner { signer, client})
+    }
+}
+
+#[async_trait]
+impl CertificateProvider for CustomSigner {
+    #[instrument(skip_all)]
+    async fn fetch_certificate(&self, id: &Identity) -> Result<tls::Certs, Error> {
+        let client = self.client.clone();
+        let csr: Api<CertificateSigningRequest> = Api::all(client.clone());
+        let pp = PostParams::default();
+        let cs = tls::CsrOptions {
+            san: id.to_string(),
+        }
+        .generate()?;
+        let cert: Vec<u8> = cs.csr;
+        let pkey = cs.pkey;
+        let csr_name = generate_csr_name();
+        let stored_csr_name = csr_name.clone();
+        let signer_name = &self.signer;
+        let mut usage = Vec::new();
+        usage.push("client auth".to_string());
+
+        let obj_metadata = k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta{
+            name: Some(csr_name),
+            ..Default::default()
+        };
+        let csr_spec = CertificateSigningRequestSpec{
+            expiration_seconds:Some(86400),
+            request:k8s_openapi::ByteString(cert),
+            signer_name:signer_name.as_ref().unwrap().to_string(),
+            usages:Some(usage),
+            ..Default::default()
+        };
+
+        let csr_req = CertificateSigningRequest {
+            metadata: obj_metadata,
+            spec: csr_spec,
+            ..Default::default()
+        };
+
+        // creating CSR
+        match csr.create(&pp, &csr_req).await {
+            Ok(o) => {
+                let name = o.metadata.name.unwrap_or_default();
+                assert_eq!(csr_req.metadata.name.unwrap_or_default(), name);
+            }
+            Err(kube::Error::Api(ae)) => assert_eq!(ae.code, 409),
```

Comment:
```suggestion
            Err(kube::Error::Api(ae)) => assert_eq!(ae.code, 409),
```

This should go away, it can fail for any other reason as well. This would panic